Test/FMI3/fmi3_import_type_definitions_test.cpp

```diff
@@ -0,0 +1,1393 @@
+    size_t maxSize;
+    size_t maxSizeDefault = 0;
+
+    //--------------------------------------------------------------------------
+    // Test attributes for TypeDefinitions
+    //--------------------------------------------------------------------------
+    fmi3_import_type_definitions_t* tds; 
+    fmi3_import_binary_typedef_t *tdMinimal, *tdMimeType, *tdAllAttr;
+
+    tds = fmi3_import_get_type_definitions(xml);
+    REQUIRE(tds != nullptr);
+    
+    tdMinimal  = fmi3_import_get_type_as_binary(get_typedef_by_name(tds, "td_minimal"));
+    tdMimeType = fmi3_import_get_type_as_binary(get_typedef_by_name(tds, "td_mimeType"));
+    tdAllAttr  = fmi3_import_get_type_as_binary(get_typedef_by_name(tds, "td_allAttr"));
+    REQUIRE(tdMinimal  != nullptr);
+    REQUIRE(tdMimeType != nullptr);
+    REQUIRE(tdAllAttr  != nullptr);
+    
+    SECTION("Minimal typedef") {
+        mimeType = fmi3_import_get_binary_type_mime_type(tdMinimal);
+        maxSize  = fmi3_import_get_binary_type_max_size(tdMinimal);
+        REQUIRE(strcmp(mimeType, mimeTypeDefault) == 0);
+        REQUIRE(maxSize == maxSizeDefault);
+    }
+    SECTION("Typedef that sets mimeType") {
+        mimeType = fmi3_import_get_binary_type_mime_type(tdMimeType);
+        maxSize  = fmi3_import_get_binary_type_max_size(tdMimeType);
+        REQUIRE(strcmp(mimeType, "mt0") == 0);
+        REQUIRE(maxSize == maxSizeDefault);
+    }
+    SECTION("Typedef that sets both mimeType and maxSize") {
+        mimeType = fmi3_import_get_binary_type_mime_type(tdAllAttr);
+        maxSize  = fmi3_import_get_binary_type_max_size(tdAllAttr);
+        REQUIRE(strcmp(mimeType, "mt0") == 0);
+        REQUIRE(maxSize == 999);
+    }
+    
+    //---------------------------------------------------------------------------------
+    // Test attribute for variables with declaredType, but without override on Variable
+    //---------------------------------------------------------------------------------
+
+    fmi3_import_variable_t* v;
+    fmi3_import_binary_variable_t* bv;
+    
+    SECTION("Variable without typedef") {
+        v = fmi3_import_get_variable_by_name(xml, "var_default_type_1");
+        bv = fmi3_import_get_variable_as_binary(v);
+        mimeType = fmi3_import_get_binary_variable_mime_type(bv);
+        maxSize  = fmi3_import_get_binary_variable_max_size(bv);
+        REQUIRE(fmi3_import_get_variable_declared_type(v) == nullptr);
+        REQUIRE(strcmp(mimeType, mimeTypeDefault) == 0);
+        REQUIRE(maxSize == maxSizeDefault);
+    }
+    SECTION("Variable inheriting minimal typedef") {
+        v = fmi3_import_get_variable_by_name(xml, "var_inh_minimal_1");
+        bv = fmi3_import_get_variable_as_binary(v);
+        mimeType = fmi3_import_get_binary_variable_mime_type(bv);
+        maxSize  = fmi3_import_get_binary_variable_max_size(bv);
+        REQUIRE(fmi3_import_get_variable_declared_type(v) == (void*)tdMinimal);
```

Comment:
What is the purpose of the (void*) here? This appears in a bunch of places and I don't quite understand it, since the return type of fmi3_import_get_variable_declared_type matches up with the type of here tdMinimal.

Reply:
No, they are different. LHS is `fmi3_import_variable_typedef_t*` and RHS is `fmi3_import_binary_typedef_t*`. In this specific test I could've used the API call that converts to the specific type's typedef, but I kind of took a shortcut since all that functions just do a typecast.

You can assign or compare a void* with any kind of void pointer, so that's a short way of not needing to write the whole super long type name everywhere.

On the places where it's assignment, casting to void* is the same as casting to whatever the type is of the declared variable. Just less text, and no need to change the typecast if the declared type changes.

Reply:
Ok, thanks for clarifying.